pubsub/gcloud/aio/pubsub/subscriber.py

```diff
@@ -0,0 +1,222 @@
+                await self.refresh()
+            return self.ack_deadline
+
+        async def refresh(self) -> None:
+            try:
+                sub = await self.subscriber_client.get_subscription(
+                    self.subscription)
+                self.ack_deadline = float(sub['ackDeadlineSeconds'])
+            except Exception as e:
+                log.warning(
+                    'Failed to refresh ackDeadlineSeconds value', exc_info=e)
+            self.last_refresh = time.perf_counter()
+
+        def cache_outdated(self) -> bool:
+            if (time.perf_counter() - self.last_refresh) > self.cache_timeout:
+                return True
+            return False
+
+    async def acker(subscription: str,
+                    ack_queue: 'asyncio.Queue[str]',
+                    subscriber_client: 'SubscriberClient',
+                    ack_window: float,
+                    metrics_client: MetricsAgent) -> None:
+        ack_ids: List[str] = []
+        while True:
+            time_budget = ack_window
+            if not ack_ids:
+                ack_ids.append(await ack_queue.get())
+                ack_queue.task_done()
+
+            while time_budget > 0:
+                start = time.perf_counter()
+                try:
+                    ack_id = await asyncio.wait_for(
+                        ack_queue.get(), timeout=time_budget)
+                    ack_ids.append(ack_id)
+                    ack_queue.task_done()
+                except asyncio.TimeoutError:
+                    break
+                time_budget -= (time.perf_counter() - start)
+
+            try:
+                await subscriber_client.acknowledge(subscription,
+                                                    ack_ids=ack_ids)
+            except Exception as e:
+                log.warning(
+                    'Ack request failed, better luck next batch', exc_info=e)
+                continue
+
+            metrics_client.histogram('pubsub.acker.batch', len(ack_ids))
+
+            ack_ids = []
+
+    async def _fire_callback(semaphore: asyncio.Semaphore,
+                             message: SubscriberMessage,
+                             callback: ApplicationHandler,
+                             ack_queue: 'asyncio.Queue[str]',
+                             metrics_client: MetricsAgent
+                             ) -> None:
+        try:
+            start = time.perf_counter()
+            await callback(message)
+            await ack_queue.put(message.ack_id)
+            metrics_client.increment('pubsub.consumer.succeeded')
+            metrics_client.histogram('pubsub.consumer.latency.runtime',
+                                     time.perf_counter() - start)
+        except Exception:
+            log.exception('Application callback raised an exception')
+            metrics_client.increment('pubsub.consumer.failed')
+        finally:
+            semaphore.release()
```

Comment:
I wonder if we can make the locking and unlocking of this semaphore a responsibility of only one calling context instead of having function A lock and trust that function B unlocks. Could you set a done callback via `add_done_callback` on the `_fire_callback` future that runs `semaphore.release()`? This is a kind of "don't make other people clean up your own mess" pattern (or "clean up your own memory" pattern). On the other hand what you have is fine and in all reality is probably more performant.

Reply:
Hm, that makes sense to me. Something like this?
```
task = asyncio.ensure_future(_execute_callback(
                message,
                callback,
                ack_queue,
                metrics_client,
            ))
task.add_done_callback(lambda x: semaphore.release())
```